src/net_processing.cpp

```diff
@@ -3305,6 +3314,43 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
     return true;
 }
 
+void CheckLocalServices(ServiceFlags& nLocalServices) {
+    if (fPruneMode) {
+        LOCK(cs_main);
+
+        // make sure NODE_NETWORK is disabled
+        nLocalServices = ServiceFlags(nLocalServices & ~NODE_NETWORK);
+
+        // don't signal NODE_NETWORK_LIMITED during initial block download
+        if (IsInitialBlockDownload()) {
+            return;
+        }
+
+        // check how many blocks we have available (backwards from tip)
+        // don't go futher down then chainActive.Height()-NODE_NETWORK_LIMITED
+        CBlockIndex *block = chainActive.Tip();
+        unsigned int chainHeight = (unsigned int) chainActive.Height();
+        while (block && block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA) && block->pprev->nTx > 0 && chainHeight - block->nHeight < (NODE_NETWORK_LIMITED_MIN_BLOCKS + 1)) {
+            block = block->pprev;
+        }
+        unsigned int height = (unsigned int) block->nHeight;
+
+        // set NODE_NETWORK_LIMITED flag based on the available blocks
+        if (chainHeight - height >= NODE_NETWORK_LIMITED_MIN_BLOCKS) {
+            nLocalServices = ServiceFlags(nLocalServices | NODE_NETWORK_LIMITED);
+        }
+
+        // don't set NODE_NETWORK if manual pruning is enabled and no blocks have been pruned.
+        // A manual prune will be very likely
+    }
+    else {
+        // currently, the prune mode can't be changed during runtime
+        // implement the service bit switch for possible future flexibility
+        nLocalServices = ServiceFlags(nLocalServices | NODE_NETWORK);
+        nLocalServices = ServiceFlags(nLocalServices & ~NODE_NETWORK_LIMITED);
```

Comment:
Why? The BIP seems somewhat unclear here, but I'd stronlgy prefer that NODE_NETWORK_LIMITED simply be always set to indicate that we can always provide the most recent 288 blocks, irrespective of if we're pruned or not.